File: src/main/java/org/taktik/icure/entities/embed/Permission.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.entities.embed;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.taktik.icure.constants.Permissions;
import org.taktik.icure.security.PermissionSetIdentifier;

import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
public class Permission implements Cloneable, Serializable {
	private static final long serialVersionUID = 1L;

	protected int grant;
	protected int revoke;
	protected Set<PermissionCriterion> criteria = new HashSet<>();

    public static Permission granted(Permissions.Type... permissions) {
        Permission p = new Permission();
        for(Permissions.Type pt:permissions) { p.grant(pt); }

        return  p;
    }

	public void grant(Permissions.Type permissionType) {
		// Grant TYPE
		grant |= permissionType.getBitValue();

		// Make sure TYPE is not revoked
		revoke &= ~permissionType.getBitValue();
	}

	public void revoke(Permissions.Type permissionType) {
		// Revoke TYPE
		revoke |= permissionType.getBitValue();

		// Make sure TYPE is not granted
		grant &= ~permissionType.getBitValue();
	}

	public void keep(Permissions.Type permissionType) {
		// Make sure TYPE is not granted
		grant &= ~permissionType.getBitValue();

		// Make sure TYPE is not revoked
		revoke &= ~permissionType.getBitValue();
	}

	@JsonIgnore
    public boolean isGranted(Permissions.Type permissionType) {
		return permissionType.isEnabled(grant);
	}

    @JsonIgnore
    public boolean isRevoked(Permissions.Type permissionType) {
		return permissionType.isEnabled(revoke);
	}

    @JsonIgnore
    public boolean isKept(Permissions.Type permissionType) {
		return !isGranted(permissionType) && !isRevoked(permissionType);
	}

	public int getGrant() {
		int grant = this.grant;

		// Remove invalid grant TYPE
		for (Permissions.Type permissionType : Permissions.Type.values()) {
			if (!canBeUsedWith(permissionType)) {
				grant &= ~permissionType.getBitValue();
			}
		}

		return grant;
	}

	public int getRevoke() {
		int revoke = this.revoke;

		// Remove invalid revoke TYPE
		for (Permissions.Type permissionType : Permissions.Type.values()) {
			if (!canBeUsedWith(permissionType)) {
				revoke &= ~permissionType.getBitValue();
			}
		}

		return revoke;
	}

    protected boolean canBeUsedWith(Permissions.Type permissionType) {
		if (getCriteria() != null && !getCriteria().isEmpty()) {
			for (PermissionCriterion criterion : getCriteria()) {
				if (!permissionType.isCriterionTypeSupported(criterion.getType())) {
					return false;
				}
			}

			return true;
		}

		return permissionType.isNoCriterionSupported();
	}

	public Set<PermissionCriterion> getCriteria() {
		return criteria;
	}

	public void setGrant(int grant) {
		this.grant = grant;
	}

	public void setRevoke(int revoke) {
		this.revoke = revoke;
	}

    @JsonIgnore
    public boolean isUseless() {
		return (getGrant() == 0 && getRevoke() == 0);
	}

	public boolean conflictWith(Permission permission) {
		if (hasSameCriteriaAs(permission)) {
			return ((getGrant() & permission.getRevoke()) != 0 || (getRevoke() & permission.getGrant()) != 0);
		}

		return false;
	}

    public boolean hasNoCriteria() {
		return (criteria == null || criteria.isEmpty());
	}

    public boolean hasSameCriteriaAs(Permission permission) {
		if (criteria == null) {
			return permission.getCriteria() == null;
		}
		return permission.getCriteria() != null && criteria.equals(permission.getCriteria());
	}

	public boolean superScope(Permission lookup) {
		// TODO improve this
		return (hasNoCriteria() && !lookup.hasNoCriteria());
	}

	public void setCriteria(Set<PermissionCriterion> value) {
		this.criteria = value;
	}

	public void addToCriteria(PermissionCriterion value) {
		this.criteria.add(value);
	}

	public void removeFromCriteria(PermissionCriterion value) {
		this.criteria.remove(value);
	}

    public int countMatchedCriteria(PermissionSetIdentifier permissionSetIdentifier, String dataCreationUserId, String dataModificationUserId,
                                    String patientCreationUserId, String patientModificationUserId,
                                    String patientReferenceHcParty, Set<String> patientHcPartiesTeam,
                                    Permissions.CriterionDataType matchedDataType, String matchedPatientStatus) {
        int matched = 0;

        if (criteria != null) {
            for (PermissionCriterion criterion : criteria) {
                if (criterion.match(permissionSetIdentifier, dataCreationUserId, dataModificationUserId, patientCreationUserId, patientModificationUserId, patientReferenceHcParty, patientHcPartiesTeam, matchedDataType, matchedPatientStatus)) {
                    matched++;
                } else {
                    return -1;
                }
            }
        }

        return matched;
    }

	@Override
	public Permission clone() {
		Permission clone = new Permission();
		clone.setGrant(getGrant());
		clone.setRevoke(getRevoke());
		if (criteria != null) {
			for (PermissionCriterion criterion : criteria) {
				PermissionCriterion criterionClone = criterion.clone();
				clone.addToCriteria(criterionClone);
			}
		}
		return clone;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + grant;
		result = prime * result + revoke;
		result = prime * result + ((criteria == null) ? 0 : criteria.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Permission other = (Permission) obj;
		if (grant != other.grant)
			return false;
		if (revoke != other.revoke)
			return false;
		if (criteria == null) {
			if (other.criteria != null)
				return false;
		} else if (!criteria.equals(other.criteria))
			return false;
		return true;
	}
}

File: src/main/kotlin/org/taktik/icure/entities/embed/Permission.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.entities.embed

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.github.pozo.KotlinBuilder
import org.taktik.icure.constants.Permissions
import org.taktik.icure.constants.Permissions.CriterionDataType
import org.taktik.icure.security.PermissionSetIdentifier
import java.io.Serializable
import java.util.*

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonIgnoreProperties(ignoreUnknown = true)
@KotlinBuilder
data class Permission(var grant: Int = 0, var revoke: Int = 0, var criteria: MutableSet<PermissionCriterion?>? = HashSet()) : Cloneable, Serializable {

    fun grant(permissionType: Permissions.Type) {
        // Grant TYPE
        grant = grant or permissionType.bitValue

        // Make sure TYPE is not revoked
        revoke = revoke and permissionType.bitValue.inv()
    }

    fun revoke(permissionType: Permissions.Type) {
        // Revoke TYPE
        revoke = revoke or permissionType.bitValue

        // Make sure TYPE is not granted
        grant = grant and permissionType.bitValue.inv()
    }

    fun keep(permissionType: Permissions.Type) {
        // Make sure TYPE is not granted
        grant = grant and permissionType.bitValue.inv()

        // Make sure TYPE is not revoked
        revoke = revoke and permissionType.bitValue.inv()
    }

    @JsonIgnore
    fun isGranted(permissionType: Permissions.Type): Boolean {
        return permissionType.isEnabled(grant)
    }

    @JsonIgnore
    fun isRevoked(permissionType: Permissions.Type): Boolean {
        return permissionType.isEnabled(revoke)
    }

    @JsonIgnore
    fun isKept(permissionType: Permissions.Type): Boolean {
        return !isGranted(permissionType) && !isRevoked(permissionType)
    }

    protected fun canBeUsedWith(permissionType: Permissions.Type): Boolean {
        if (criteria != null && criteria!!.isNotEmpty()) {
            for (criterion in criteria!!) {
                if (!permissionType.isCriterionTypeSupported(criterion!!.getType())) {
                    return false
                }
            }
            return true
        }
        return permissionType.isNoCriterionSupported
    }


    @get:JsonIgnore
    val isUseless: Boolean
        get() = grant == 0 && revoke == 0

    fun conflictWith(permission: Permission): Boolean {
        return if (hasSameCriteriaAs(permission)) {
            grant and permission.revoke != 0 || revoke and permission.grant != 0
        } else false
    }

    fun hasNoCriteria(): Boolean {
        return criteria == null || criteria!!.isEmpty()
    }

    fun hasSameCriteriaAs(permission: Permission): Boolean {
        return if (criteria == null) {
            permission.criteria == null
        } else permission.criteria != null && criteria == permission.criteria
    }

    fun superScope(lookup: Permission): Boolean {
        // TODO improve this
        return hasNoCriteria() && !lookup.hasNoCriteria()
    }

    fun addToCriteria(value: PermissionCriterion?) {
        criteria!!.add(value)
    }

    fun removeFromCriteria(value: PermissionCriterion?) {
        criteria!!.remove(value)
    }

    fun countMatchedCriteria(permissionSetIdentifier: PermissionSetIdentifier, dataCreationUserId: String?, dataModificationUserId: String?,
                             patientCreationUserId: String?, patientModificationUserId: String?,
                             patientReferenceHcParty: String?, patientHcPartiesTeam: Set<String?>?,
                             matchedDataType: CriterionDataType?, matchedPatientStatus: String?): Int {
        var matched = 0
        if (criteria != null) {
            for (criterion in criteria!!) {
                if (criterion!!.match(permissionSetIdentifier, dataCreationUserId, dataModificationUserId, patientCreationUserId, patientModificationUserId, patientReferenceHcParty, patientHcPartiesTeam, matchedDataType, matchedPatientStatus)) {
                    matched++
                } else {
                    return -1
                }
            }
        }
        return matched
    }

    companion object {
        fun granted(vararg permissions: Permissions.Type): Permission {
            val p = Permission()
            for (pt in permissions) {
                p.grant(pt)
            }
            return p
        }
    }
}
